File: Sources/AvasiaEngine/Model/Room.swift
import Foundation

/// What a turn produces: lines to print, where to go next, and any semantic
/// events for the achievement tracker.
public struct TurnResult: Sendable {
    public var lines: [StyledLine]
    public var transition: Transition
    public var events: [GameEvent]

    public init(_ lines: [StyledLine] = [], _ transition: Transition = .stay, events: [GameEvent] = []) {
        self.lines = lines
        self.transition = transition
        self.events = events
    }
}

/// The outcome of a command, applied by `GameEngine`.
public enum Transition: Sendable, Equatable {
    case stay                       // remain in the current room
    case move(RoomID)               // go to another room
    case death(DeathCause)          // lethal — increments deathCount, offers restart
    case win                        // reached the canonical ending
}

/// A room knows how to describe itself and handle player input. Implementations
/// are pure: they read and mutate `GameState` but perform no I/O. This makes the
/// whole world unit-testable by scripting input sequences.
public protocol RoomScript: Sendable {
    var id: RoomID { get }

    /// Which parser mode this room uses (most are `.normalized`; the cave hub,
    /// trading post, bridge and yes/no prompts use `.raw` — ENGINE_SPEC §B.4).
    var parseMode: Parser.Mode { get }

    /// The room's description, possibly branching on state (e.g. escort mode).
    func describe(_ state: GameState) -> [StyledLine]

    /// Handle one parsed command.
    func handle(_ input: ParsedInput, _ state: inout GameState) -> TurnResult
}

public extension RoomScript {
    var parseMode: Parser.Mode { .normalized }
}

// Shared synonym lists matching the original's per-room dispatch lists.
public enum Verb {
    public static let north = ["NORTH"]
    public static let east  = ["EAST"]
    public static let west  = ["WEST"]
    public static let south = ["SOUTH"]
    public static let back  = ["LEAVE", "BACK", "EXIT", "RETURN", "SOUTH"]
    public static let up    = ["UP", "STAIRS", "UPSTAIRS"]
    public static let down  = ["DOWN", "DOWNSTAIRS"]
    public static let left  = ["LEFT"]
    public static let right = ["RIGHT"]
    public static let look  = ["LOOK", "EXAMINE", "INSPECT"]
    public static let talk  = ["TALK", "SPEAK"]
    public static let take  = ["TAKE", "GRAB", "STEAL", "GET", "PICK"]
}

File: Avasia-SoC/Room/Room.py
from Logic.util import containsAny, set_color_to
import Logic.config as config


class Room:
    def __init__(self, name, des, id, directions, on_enter=None):
        self.name = name
        self.des = des
        self.id = id
        self.directions = directions
        self.on_enter = on_enter

    def event(self):
        # This is a linking room
        if self.on_enter is None:
            self.print_name()
            print(self.des)
        else:
            # Logic Room
            return self.on_enter()

    def print_name(self):
        set_color_to(config.room_color)
        print("\n>>>" + self.name + "<<<\n\n")
        set_color_to(config.base_color)

    def direction(self, command):
        north = ["NORTH"]
        east = ["EAST"]
        west = ["WEST"]
        south = ["SOUTH"]
        northwest = ["NORTHWEST"]
        northeast = ["NORTHEAST"]
        d = ["DOWN", "DOWNSTAIRS"]
        up = ["UP", "UPSTAIRS"]
        left = ["LEFT"]
        r = ["RIGHT"]
        if containsAny(command, north):
            if "N" in self.directions:
                config.current_room_id = self.directions["N"]
        if containsAny(command, east):
            if "E" in self.directions:
                config.current_room_id = self.directions["E"]
        if containsAny(command, west):
            if "W" in self.directions:
                config.current_room_id = self.directions["W"]
        if containsAny(command, south):
            if "S" in self.directions:
                config.current_room_id = self.directions["S"]
        if containsAny(command, northwest):
            if "NW" in self.directions:
                config.current_room_id = self.directions["NW"]
        if containsAny(command, northeast):
            if "NE" in self.directions:
                config.current_room_id = self.directions["NE"]
        if containsAny(command, left):
            if "L" in self.directions:
                config.current_room_id = self.directions["L"]
        if containsAny(command, r):
            if "R" in self.directions:
                config.current_room_id = self.directions["R"]
        if containsAny(command, up):
            if "U" in self.directions:
                config.current_room_id = self.directions["U"]
        if containsAny(command, d):
            if "D" in self.directions:
                config.current_room_id = self.directions["D"]


